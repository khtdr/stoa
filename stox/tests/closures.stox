#!/usr/bin/env stox

var counter = "outer-scope";
print counter;

var inc;
{
  var counter = 0;
  print counter;
  fun incr() { counter = counter + 1; return counter; }
  inc = incr;
}
print inc();
print inc();
print inc();

print counter;

fun makeDec(i) {
  fun dec() { i =i-1; return i; }
  return dec;
}

var i = 12;
var dec = makeDec(100);
print dec();
print dec();
print i;

fun makeCounter(start) {
  var n = start;
  return fun() {
    n = n+1;
    return n;
  };
}

var c = makeCounter(212);
print c();

// print makeCounter(212)();

// need to peek ahead?
// fun() {};
