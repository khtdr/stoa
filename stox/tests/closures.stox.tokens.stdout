VAR[3,1]
IDENTIFIER[3,5]
EQUAL[3,13]
STRING[3,15]
SEMICOLON[3,28]
PRINT[4,1]
IDENTIFIER[4,7]
SEMICOLON[4,14]
VAR[6,1]
IDENTIFIER[6,5]
SEMICOLON[6,8]
LEFT_CURL[7,1]
VAR[8,3]
IDENTIFIER[8,7]
EQUAL[8,15]
NUMBER[8,17]
SEMICOLON[8,18]
PRINT[9,3]
IDENTIFIER[9,9]
SEMICOLON[9,16]
FUN[10,3]
IDENTIFIER[10,7]
LEFT_PAREN[10,11]
RIGHT_PAREN[10,12]
LEFT_CURL[10,14]
IDENTIFIER[10,16]
EQUAL[10,24]
IDENTIFIER[10,26]
PLUS[10,34]
NUMBER[10,36]
SEMICOLON[10,37]
RETURN[10,39]
IDENTIFIER[10,46]
SEMICOLON[10,53]
RIGHT_CURL[10,55]
IDENTIFIER[11,3]
EQUAL[11,7]
IDENTIFIER[11,9]
SEMICOLON[11,13]
RIGHT_CURL[12,1]
PRINT[13,1]
IDENTIFIER[13,7]
LEFT_PAREN[13,10]
RIGHT_PAREN[13,11]
SEMICOLON[13,12]
PRINT[14,1]
IDENTIFIER[14,7]
LEFT_PAREN[14,10]
RIGHT_PAREN[14,11]
SEMICOLON[14,12]
PRINT[15,1]
IDENTIFIER[15,7]
LEFT_PAREN[15,10]
RIGHT_PAREN[15,11]
SEMICOLON[15,12]
PRINT[17,1]
IDENTIFIER[17,7]
SEMICOLON[17,14]
FUN[19,1]
IDENTIFIER[19,5]
LEFT_PAREN[19,12]
IDENTIFIER[19,13]
RIGHT_PAREN[19,14]
LEFT_CURL[19,16]
FUN[20,3]
IDENTIFIER[20,7]
LEFT_PAREN[20,10]
RIGHT_PAREN[20,11]
LEFT_CURL[20,13]
IDENTIFIER[20,15]
EQUAL[20,17]
IDENTIFIER[20,18]
DASH[20,19]
NUMBER[20,20]
SEMICOLON[20,21]
RETURN[20,23]
IDENTIFIER[20,30]
SEMICOLON[20,31]
RIGHT_CURL[20,33]
RETURN[21,3]
IDENTIFIER[21,10]
SEMICOLON[21,13]
RIGHT_CURL[22,1]
VAR[24,1]
IDENTIFIER[24,5]
EQUAL[24,7]
NUMBER[24,9]
SEMICOLON[24,11]
VAR[25,1]
IDENTIFIER[25,5]
EQUAL[25,9]
IDENTIFIER[25,11]
LEFT_PAREN[25,18]
NUMBER[25,19]
RIGHT_PAREN[25,22]
SEMICOLON[25,23]
PRINT[26,1]
IDENTIFIER[26,7]
LEFT_PAREN[26,10]
RIGHT_PAREN[26,11]
SEMICOLON[26,12]
PRINT[27,1]
IDENTIFIER[27,7]
LEFT_PAREN[27,10]
RIGHT_PAREN[27,11]
SEMICOLON[27,12]
PRINT[28,1]
IDENTIFIER[28,7]
SEMICOLON[28,8]
FUN[30,1]
IDENTIFIER[30,5]
LEFT_PAREN[30,16]
IDENTIFIER[30,17]
RIGHT_PAREN[30,22]
LEFT_CURL[30,24]
VAR[31,3]
IDENTIFIER[31,7]
EQUAL[31,9]
IDENTIFIER[31,11]
SEMICOLON[31,16]
RETURN[32,3]
FUN[32,10]
LEFT_PAREN[32,13]
RIGHT_PAREN[32,14]
LEFT_CURL[32,16]
IDENTIFIER[33,5]
EQUAL[33,7]
IDENTIFIER[33,9]
PLUS[33,10]
NUMBER[33,11]
SEMICOLON[33,12]
RETURN[34,5]
IDENTIFIER[34,12]
SEMICOLON[34,13]
RIGHT_CURL[35,3]
SEMICOLON[35,4]
RIGHT_CURL[36,1]
VAR[38,1]
IDENTIFIER[38,5]
EQUAL[38,7]
IDENTIFIER[38,9]
LEFT_PAREN[38,20]
NUMBER[38,21]
RIGHT_PAREN[38,24]
SEMICOLON[38,25]
PRINT[39,1]
IDENTIFIER[39,7]
LEFT_PAREN[39,8]
RIGHT_PAREN[39,9]
SEMICOLON[39,10]
